syntex_syntax/src/ext/decorator.rs

```diff
@@ -149,23 +160,34 @@ fn expand_decorators_2(
                 fld.cx.meta_list(mitem.span, mname.as_str(), not_handled));
             new_attrs.push(derive);
         }
+        item
     } else {
-        let handled = expand_decorators_3(item, attr, fld, decorator_items, mname);
+        let (handled, out) = expand_3(item, attr, fld, out_items, mname);
         if !handled {
             new_attrs.push((*attr).clone());
         }
+        out
     }
 }
 
-// Responsible for expanding attributes that match a MultiDecorator registered
-// in the syntax_env. Returns whether the given attribute was handled.
-fn expand_decorators_3(
-    item: &Annotatable,
+// Responsible for expanding attributes that match a MultiDecorator or
+// MultiModifier registered in the syntax_env. Returns whether the given
+// attribute was handled, along with the item to continue processing.
+//
+// Syntex supports only a special case of MultiModifier - those that produce
+// exactly one output. If a MultiModifier produces zero or more than one output
+// this function panics. The problematic case we cannot support is:
+//
+//     #[decorator] // not registered with Syntex
+//     #[modifier] // registered
+//     struct A;
+fn expand_3(
+    item: Annotatable,
     attr: &ast::Attribute,
     fld: &mut MacroExpander,
-    decorator_items: &mut SmallVector<Annotatable>,
+    out_items: &mut SmallVector<Annotatable>,
     mname: Name,
-) -> bool {
+) -> (bool, Annotatable) {
```

Comment:
not really important, but isn't this basically a `Result<Annotatable, Annotatable>`, where the `Ok` means "handled", and the `Err` means "not handled, here's your `Annotatable` back"